File: Scrabble.py
print "Welcome to the Scrabble Calculator"

""" functions """
# word list changes values 
def userEnterWords(numPlayers):
    for i in range (int(numPlayers)):
        word[i] = raw_input("Enter a word.")
    print word

# for every character in the list, send it into a value returning function
# that returns a value that will get added to the total
def addToTotal(name, word, total, numPlayers):
    for i in range(int(numPlayers)):
        for character in word[i]:
            total[i] += returnPoints(character)

def printCurrentScore(name, total, numPlayers):
    for i in range(int(numPlayers)):
        print name[i], "has", total[i], "."
        
# long ifelse statement that give scrabble values
def returnPoints(x):
    if x == 'e' or x == 'a' or x == 'i' or x == 'o' or x == 'n':
        return 1
    if x == 'r' or x == 't' or x == 'l' or x == 's' or x == 'u':
        return 1
    if x == 'd' or x == 'g':
        return 2
    if x == 'b' or x == 'c' or x == 'm' or x == 'p':
        return 3
    if x == 'f' or x == 'h' or x == 'v' or x == 'w' or x == 'y':
        return 4
    if x == 'k':
        return 5
    if x == 'j' or x == 'x':
        return 8
    if x == 'q' or x == 'z':
        return 10
    if x == '_':
        return 0
    if x == 'Q' or x == 'Z':
        return 20
    if x == 'J' or x == 'X':
        return 16
    if x == 'K':
        return 10
    if x == 'F' or x == 'H' or x == 'V' or x == 'W' or x == 'Y':
        return 8
    if x == 'B' or x == 'C' or x == 'M' or x == 'P':
        return 6
    if x == 'D' or x == 'G':
        return 4
    if x == 'T' or x == 'L' or x == 'S' or x == 'U':
        return 2
    if x == 'E' or x == 'A' or x == 'I' or x == 'O' or x == 'N' or x == 'R':
        return 2

# returns true if users enter yes, ends the game
def usersAreQuitting():
    userChoice = raw_input("Is the game done? (y/n)")
    if userChoice == 'y':
        return True
    elif userChoice == 'n':
        return False
    else:
        userChoice == raw_input("You didn't enter anything. Enter again.")
        while userChoice != 'y' or userChoice != 'n':
            userChoice == raw_input("You didn't enter anything. Enter again.")

# declare the winner 
def declareWinner(name, total, numPlayers):
    highest = total[0]
    counter = 0 # checks for draws
    for i in range(1,int(numPlayers)):
        if highest < total[i]:
            highest = total[i]
    for i in range(int(numPlayers)):
        if highest == total[i]:
            print name[i], "has the highest score."
            counter = counter + 1
    if counter > 1:
        print "There is more than one winner."

""" Main """
numPlayers = raw_input("How many players are playing? ")
name = [] #list declaration
word = ["nothing", "nothing", "nothing", "nothing"] #list of words
total = [0, 0, 0, 0] #everyone starts with 0 points

for i in range(int(numPlayers)):# int(numPlayers) gets the interger value of raw_input
    nameVariable = raw_input("Enter your name. ")
    name.append(nameVariable)

print name

while True:
    userEnterWords(numPlayers)
    addToTotal(name, word, total, numPlayers) #processes words to points and adds them to totals
    printCurrentScore(name, total, numPlayers)
    if(usersAreQuitting()):
        break

declareWinner(name, total, numPlayers)

File: Scrabble.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>

using namespace std;

struct typep {
	string name;
	string words;
	int score;
};
/*  welcomes the user and asks them how many players are playing */
void welcome(int& amountOfPlayers);
/*  user enters in their names */
void getNames(vector<typep>& player, int& amountOfPlayers);
/*  function where user enters in their words to be scored */
void userAddsWords(vector<typep>& player, int& amountOfPlayers);
/*  uses the scrabble function and adds each score to the vector holding totals */
void calculateWords(vector<typep>& player, int& amountOfPlayers);
/* prints the score after each player is finished with their words */
void sayTheScore(vector<typep>& player, int& amountOfPlayers);
/*  winner is declared, after using the checkForWinner function */
void declareWinner(vector<typep>& player, int& amountOfPlayers);
/*  function that holds all the value of each character */
void Scrabble(char aa, int& add);
/*  will loop through the function for as long as game continues */
void loop(vector<typep>& player, int& amountOfPlayers);
/*  checks for the winner */
bool checkForWinner(int x, vector<typep>& player, int& amountOfPlayers);

int main() {
	int amountOfPlayers; 
	welcome(amountOfPlayers); /*  determines amount of players playing the game */
	vector<typep> player(amountOfPlayers); 
	getNames(player, amountOfPlayers);
    loop(player, amountOfPlayers);
	declareWinner(player, amountOfPlayers);
	return 0;
}

void welcome(int& amountOfPlayers) {
	cout << "Welcome to the Scrabble Calculator." << endl;
	cout << "How many people are playing?" << endl;
	cin >> amountOfPlayers;
	if(amountOfPlayers>4) {
		cout << "4 is the max for Scrabble. Enter again." << endl;
		cin >> amountOfPlayers;
		while(!(amountOfPlayers<4)) {
			cout << "Enter again." << endl;
			cin >> amountOfPlayers;
		}
	}
}
	
void getNames(vector<typep>& player, int& amountOfPlayers) {
	int x = 1;
	for(int i=0; i<amountOfPlayers; i++) {
		cout << "Player " << x << " ?" << endl;
		string name;
		cin >> name;
		player[i].name = name;
		x += 1;
	}
	cout << "The players are: " << endl;
	for(int i=0; i<amountOfPlayers; i++) {
		cout << player[i].name;
		cout << (i<amountOfPlayers-1 ? ", " : ".") << endl;
	}
}

void userAddsWords(vector<typep>& player, int& amountOfPlayers) {
	cout << endl;
	string input;
	for(int i=0; i<amountOfPlayers; i++) {
		cout << player[i].name << " 's turn." << endl;
		cin >> input;
		player[i].words = input;
	}
	/*check
	for(int i=0; i<amountOfPlayers; i++) {
		cout << player[i].name << player[i].words << endl;
	}*/
}

void calculateWords(vector<typep>& player, int& amountOfPlayers) {
	vector<int> toty(amountOfPlayers);
	for(int i=0; i<amountOfPlayers; i++) {
		toty[i] = 0;
	}
	for(int i=0; i<amountOfPlayers; i++) {
		int totalFromScrabble =0;
		for(int j=0; j<player[i].words.size(); j++) {
			int add =0;
			Scrabble(player[i].words[j], add);
			totalFromScrabble += add;
		}
		toty[i] += totalFromScrabble;
		player[i].score += toty[i];
	}
	/*check */
	for(int i=0; i<amountOfPlayers; i++) {
		cout << player[i].name << " has " << player[i].score; 
		cout << (i<amountOfPlayers-1 ? ", " : ". ");
	}
}

void sayTheScore(vector<typep>& player, int& amountOfPlayers) {
	for(int i=0; i<amountOfPlayers; i++) {
		cout << player[i].name << " has " << player[i].score << ". " << endl;
	}
}

void declareWinner(vector<typep>& player, int& amountOfPlayers) {
	for(int i=0; i<amountOfPlayers; i++) {
		if(checkForWinner(player[i].score, player, amountOfPlayers)) {
			cout << player[i].name << " has won." << endl;
		}
	}
}

void Scrabble(char aa, int& add) {
	switch (aa) {
		case 'e': case 'a': case 'i': case 'o': case 'n':
		case 'r': case 't': case 'l': case 's': case 'u':
			add += 1; break;
		case 'd': case 'g':
			add += 2; break;
		case 'b': case 'c': case 'm': case 'p':
			add += 3; break;
		case 'f': case 'h': case 'v': case 'w': case 'y':
			add += 4; break;
		case 'k':
			add += 5; break;
		case 'j': case 'x':
			add += 8; break;
		case 'q': case 'z':
			add += 10; break;
		case '_': 
			add += 0; break;

		//Capitals
		case 'E': case 'A': case 'I': case 'O': case 'N': case 'R': case 'T': case 'L': case 'S': case 'U':
			add += 2; break;
		case 'D': case 'G':
			add += 4; break;
		case 'B': case 'C': case 'M': case 'P':
			add += 6; break;
		case 'F': case 'H': case 'V': case 'W': case 'Y':
			add += 8; break;
		case 'K':
			add += 10; break;
		case 'J': case 'X':
			add += 16; break;
		case 'Q': case 'Z':
			add += 20; break;
		

	}
}

void loop(vector<typep>& player, int& amountOfPlayers) {
	while(true) {
		userAddsWords(player, amountOfPlayers);
		calculateWords(player, amountOfPlayers);
		cout << "Done?" << endl;
		string done;
		cin >> done;
		if(done == "done") {
			break;
		}
		sayTheScore(player, amountOfPlayers);
	}
}

bool checkForWinner(int x, vector<typep>& player, int& amountOfPlayers) {
	int highest = player[0].score;
	for(int i=1; i<amountOfPlayers; i++) {
		if(player[i].score > highest) {
			highest = player[i].score;
		}
	}
	if(x == highest) {
		return true;
	} else {
		return false;
	}
}
